### PR TITLE
Refactor Checkpoint/Restore Code into its own class

### PR DESCRIPTION
**Pull request by dsouzai**
* Refactor Checkpoint/Restore code into its own class
* Add CR Runtime Thread

Parent issue: https://github.com/eclipse-openj9/openj9/issues/18866

### CODE REVIEW

**Comment by dsouzai** [comment]
@mpirvu could you please review?

File: runtime/compiler/runtime/CRRuntime.cpp [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+   else
+      {
+      // Must wait here until the thread gets created; otherwise an early
+      // shutdown does not know whether or not to destroy the thread
+      getCRRuntimeMonitor()->enter();
+      while (getCRRuntimeThreadLifetimeState() == TR_CRRuntimeThreadLifetimeStates::CR_THR_NOT_CREATED)
+         getCRRuntimeMonitor()->wait();
+      getCRRuntimeMonitor()->exit();
+
+      // At this point the CR RUntime thread should have attached successfully
```
Comment (mpirvu) [inline runtime/compiler/runtime/CRRuntime.cpp:594]: Typo "RUntime"

File: runtime/compiler/runtime/CRRuntime.cpp [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+TR::CRRuntime::startCRRuntimeThread(J9JavaVM *javaVM)
+   {
+   PORT_ACCESS_FROM_JAVAVM(javaVM);
+
+   //256KB stack size
+   const UDATA defaultOSStackSize = javaVM->defaultOSStackSize;
+   UDATA priority = J9THREAD_PRIORITY_NORMAL;
+
+   // create the CR Runtime Thread
+   if(javaVM->internalVMFunctions->createThreadWithCategory(&_crRuntimeOSThread,
```
Comment (mpirvu) [inline runtime/compiler/runtime/CRRuntime.cpp:575]: Space needed after `if`

Comment (mpirvu) [reply runtime/compiler/runtime/CRRuntime.cpp:575]: Immediately after the thread gets created, it is possible for the crRuntime thread to execute all its code and be stuck in a busy loop in process() with CRRuntimeMonitor in hand and FAILED_TO_ATTACH state.

**Review by mpirvu — CHANGES_REQUESTED** [review]

File: runtime/compiler/runtime/CRRuntime.cpp [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] https://openjdk.org/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0-only WITH Classpath-exception-2.0 OR GPL-2.0-only WITH OpenJDK-assembly-exception-1.0
+ *******************************************************************************/
+
+#include "j9cfg.h"
+
+#if defined(J9VM_OPT_CRIU_SUPPORT)
```
[the remark this answers is not in the document]
Comment (dsouzai) [reply runtime/compiler/runtime/CRRuntime.cpp:25]: Yeah I can do that; I didn't realize that was a possibility.

**Comment by dsouzai** [comment]
@mpirvu made the requested changes. Also as we discussed offline, because the CRRuntimeThread only executes `process()` if the return value of `internalAttachCurrentThread` is `JNI_OK`, the busy loop scenario isn't possible as thing stand.

File: runtime/compiler/control/CMakeLists.txt [hunk 4]
```diff
@@ -41,3 +40,9 @@ if(J9VM_OPT_JITSERVER)
 		control/JITServerHelpers.cpp
 	)
 endif()
+
+if(J9VM_OPT_CRIU_SUPPORT)
```
Comment (mpirvu) [inline runtime/compiler/control/CMakeLists.txt:44]: Shouldn't `compiler/runtime/CRRuntime.cpp` be here as well?

Comment (mpirvu) [reply runtime/compiler/control/CMakeLists.txt:44]: Oh, it's a different directory

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by mpirvu** [comment]
jenkins test all jdk17

**Comment by mpirvu** [comment]
jenkins test sanity all jdk17

**Comment by dsouzai** [comment]
aarch64 test failure 
```
[2024-03-26T23:11:38.529Z] FAILED test targets:
[2024-03-26T23:11:38.529Z] 	cmdLineTester_criu_jitserverPostRestore_3
```
is due to
```
[2024-03-26T22:15:26.157Z] Output from test:
[2024-03-26T22:15:26.157Z]  [OUT] start running script
[2024-03-26T22:15:26.157Z]  [OUT] export GLIBC_TUNABLES=glibc.cpu.hwcaps=-XSAVEC,-XSAVE,-AVX2,-ERMS,-AVX,-AVX_Fast_Unaligned_Load
[2024-03-26T22:15:26.157Z]  [OUT] export LD_BIND_NOT=on
[2024-03-26T22:15:26.157Z]  [OUT] Starting /home/jenkins/workspace/Test_openjdk17_j9_sanity.functional_aarch64_linux_Personal_testList_0/jdkbinary/j2sdk-image/bin/jitserver -XX:JITServerPort=52366 -XX:JITServerHealthProbePort=55016 
[2024-03-26T22:15:26.157Z]  [OUT] JITSERVER DOES NOT EXIST
[2024-03-26T22:15:26.157Z]  [OUT] finished script
[2024-03-26T22:15:26.157Z]  [ERR] can't bind server address: Address already in use
[2024-03-26T22:15:26.157Z]  [ERR] Failed to open health socket on port 55016
```

**Comment by mpirvu** [comment]
jenkins test sanity alinux64 jdk17

**Comment by dsouzai** [comment]
All tests have passed; this should be good to merge now.

**Pull request merged**